File: bindings/csharp/src/STRling/Simply.cs
using System.Collections.Generic;
using Strling.Core;

namespace Strling.Simply
{
    /// <summary>
    /// Shorthand entry points for the fluent API. Short name 'S' keeps usage concise.
    /// </summary>
    public static class S
    {
        public static Pattern Literal(string text) => new Pattern(new Lit(text));

        public static Pattern Digit(int count)
        {
            return new Pattern(new Quant(new Lit("\\d"), count, count, true, false, false));
        }

        public static Pattern AnyOf(string chars)
        {
            var members = new List<ClassItem>();
            foreach (var ch in chars)
            {
                members.Add(new ClassLiteral(ch.ToString()));
            }
            return new Pattern(new CharClass(false, members));
        }

        public static Pattern Start() => new Pattern(new Anchor("Start"));

        public static Pattern End() => new Pattern(new Anchor("End"));

        public static Pattern Dot() => new Pattern(new Dot());

        public static Pattern Sequence(params Pattern[] parts)
        {
            var list = new List<Node>();
            foreach (var p in parts) list.Add(p.Node);
            return new Pattern(new Seq(list));
        }
    }
}

File: bindings/ruby/lib/strling/simply.rb
# frozen_string_literal: true

require_relative 'core/nodes'
require_relative 'ir'
require_relative 'emitters/pcre2'

module Strling
  # Fluent API for building STRling patterns
  class Simply
    attr_reader :node

    def initialize(node)
      @node = node
    end

    # Class methods (Constructors)

    def self.merge(*patterns)
      parts = patterns.map { |p| to_node(p) }
      new(Strling::Core::Seq.new(parts))
    end

    def self.capture(pattern)
      new(Strling::Core::Group.new(true, to_node(pattern)))
    end

    def self.may(pattern)
      new(Strling::Core::Quant.new(to_node(pattern), 0, 1, 'Greedy'))
    end

    def self.digit
      new(Strling::Core::Esc.new('d'))
    end

    def self.start
      new(Strling::Core::Anchor.new('Start'))
    end

    def self.end
      new(Strling::Core::Anchor.new('End'))
    end

    def self.any_of(*patterns)
      branches = patterns.map { |p| to_node(p) }
      new(Strling::Core::Alt.new(branches))
    end

    # Helper to convert input to Node
    def self.to_node(pattern)
      if pattern.is_a?(Simply)
        pattern.node
      elsif pattern.is_a?(String)
        Strling::Core::Lit.new(pattern)
      else
        raise ArgumentError, "Expected Simply instance or String, got #{pattern.class}"
      end
    end

    # Instance methods (Chainable)

    def merge(*others)
      parts = [@node] + others.map { |p| Simply.to_node(p) }
      Simply.new(Strling::Core::Seq.new(parts))
    end

    def capture
      Simply.new(Strling::Core::Group.new(true, @node))
    end

    def may
      Simply.new(Strling::Core::Quant.new(@node, 0, 1, 'Greedy'))
    end

    def any_of(*others)
      branches = [@node] + others.map { |p| Simply.to_node(p) }
      Simply.new(Strling::Core::Alt.new(branches))
    end

    def times(min, max = nil)
      max = min if max.nil?
      Simply.new(Strling::Core::Quant.new(@node, min, max, 'Greedy'))
    end

    def to_s
      ir = Strling::IR::Compiler.compile(@node)
      flags = Strling::Core::Flags.new
      Strling::Emitters::PCRE2.emit(ir, flags)
    end
  end
end
